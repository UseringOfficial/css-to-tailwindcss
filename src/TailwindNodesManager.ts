import type { ChildNode } from 'postcss';
import { type AtRule, type Node, Rule } from 'postcss';
import { isAtRuleNode } from './utils/isAtRuleNode';
import { isChildNode } from './utils/isChildNode';

export interface ResolvedTailwindNode {
  rule: Rule;
  tailwindClasses: string[];
}

export interface UnresolvedTailwindNode {
  key: string;
  rootRuleSelector?: string | null;
  originalRule: Rule;
  classesPrefix: string;
  tailwindClasses: string[];
}

export type TailwindNode = ResolvedTailwindNode | UnresolvedTailwindNode;

export function isResolvedTailwindNode(node: TailwindNode): node is ResolvedTailwindNode {
  return 'rule' in node;
}

export function isUnresolvedTailwindNode(node: TailwindNode): node is UnresolvedTailwindNode {
  return !isResolvedTailwindNode(node);
}

export class TailwindNodesManager {
  protected nodesMap: Map<string, ResolvedTailwindNode> = new Map();

  mergeNode(node: TailwindNode) {
    const nodeIsResolved = isResolvedTailwindNode(node);
    const nodeKey = nodeIsResolved ? TailwindNodesManager.convertRuleToKey(node.rule) : node.key;
    const foundNode = this.nodesMap.get(nodeKey);

    if (foundNode) {
      foundNode.tailwindClasses = foundNode.tailwindClasses.concat(
        nodeIsResolved
          ? node.tailwindClasses
          : node.tailwindClasses.map((className) => `${node.classesPrefix}${className}`),
      );

      return;
    }

    if (nodeIsResolved) {
      this.nodesMap.set(nodeKey, node);
    } else {
      let rule;

      if (node.rootRuleSelector) {
        rule = new Rule({
          selector: node.rootRuleSelector,
        });

        const rootChild = this.upToRootChild(node.originalRule);
        if (rootChild) {
          node.originalRule.root().insertBefore(rootChild, rule);
        }
      } else {
        rule = node.originalRule;
      }

      this.nodesMap.set(nodeKey, {
        rule,
        tailwindClasses: node.tailwindClasses.map((className) => `${node.classesPrefix}${className}`),
      });
    }
  }

  mergeNodes(nodes: TailwindNode[]) {
    nodes.forEach((node) => {
      this.mergeNode(node);
    });
  }

  hasNode(key: string) {
    return this.nodesMap.has(key);
  }

  getNode(key: string) {
    return this.nodesMap.get(key) || null;
  }

  getNodes() {
    return Array.from(this.nodesMap.values());
  }

  protected upToRootChild(node: Node) {
    let childNode: ChildNode | null = null;

    while (node.parent && node.parent.type !== 'root' && isChildNode(node.parent)) {
      childNode = node = node.parent;
      continue;
    }

    return childNode;
  }

  static convertRuleToKey(rule: Rule, overriddenSelector: string | null = null) {
    let currentParent = rule.parent;
    let parentKey = '';

    while (isChildNode(currentParent)) {
      parentKey += `${
        isAtRuleNode(currentParent) ? this.makeSingleAtRuleKey(currentParent) : this.makeSingleRuleKey(currentParent)
      }__`;

      currentParent = currentParent.parent;
    }

    return parentKey + (overriddenSelector == null ? rule.selector : overriddenSelector);
  }

  protected static makeSingleAtRuleKey(atRule: AtRule) {
    return `a(${atRule.name}|${atRule.params})`;
  }

  protected static makeSingleRuleKey(atRule: Rule) {
    return `r(${atRule.selector})`;
  }
}
