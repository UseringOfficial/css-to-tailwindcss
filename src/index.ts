export * from './TailwindConverter';
export * from './TailwindNodesManager';
export * from './utils/selectors';
export * from './document-converter';
