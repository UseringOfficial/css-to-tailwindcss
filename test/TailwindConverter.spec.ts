import type { TailwindConverterConfig } from '../src/TailwindConverter';
import { TailwindConverter } from '../src/TailwindConverter';
import fs from 'fs';
import path from 'path';
import { describe, expect, it } from 'vitest';
import nested from 'postcss-nested';

const complexCSS: string = fs.readFileSync(path.resolve(__dirname, './fixtures/input.css')).toString();

const simpleCSS = `
.foo {
  padding-top: 12px;
  padding-bottom: 12px;
  font-size: 12px;
  animation-delay: 200ms;
  border-right: 2px dashed;
  border: 4px solid transparent;

  &:hover {
    filter: blur(4px) brightness(0.5) sepia(100%) contrast(1) hue-rotate(30deg)
      invert(0) opacity(0.05) saturate(1.5);
    transform: translateX(12px) translateY(0.5em) translateZ(0.5rem)
      scaleY(0.725) rotate(124deg);
    font-size: 16px;
  }

  @media screen and (min-width: 768px) {
    font-weight: 600;
  }
}
`;

function createTailwindConverter(config?: Partial<TailwindConverterConfig>) {
  return new TailwindConverter({
    remInPx: 16,
    postCSSPlugins: [nested],
    tailwindConfig: {
      content: [],
      theme: {
        extend: {
          colors: {
            'custom-color': {
              100: '#123456',
              200: 'hsla(210, 100%, 51.0%, 0.016)',
              300: '#654321',
              400: 'some-invalid-color',
              gold: 'hsl(41, 28.3%, 79.8%)',
              marine: 'rgb(4, 55, 242, 0.75)',
            },
          },
          screens: {
            'custom-screen': { min: '768px', max: '1024px' },
          },
        },
        supports: {
          grid: 'display: grid',
          flex: 'display: flex',
        },
      },
    },
    ...(config || {}),
  });
}

describe('TailwindConverter', () => {
  it('should convert the simple CSS', async () => {
    const converter = createTailwindConverter();
    const converted = await converter.convertCSS(simpleCSS);

    expect(converted.convertedRoot.toString()).toMatchSnapshot();
    expect(converted.nodes).toEqual([
      {
        rule: expect.objectContaining({ selector: '.foo' }),
        tailwindClasses: [
          'text-xs',
          'py-3',
          'border-r-2',
          'border-dashed',
          'border-4',
          'border-solid',
          'border-transparent',
          'hover:blur-sm',
          'hover:brightness-50',
          'hover:sepia',
          'hover:contrast-100',
          'hover:hue-rotate-30',
          'hover:invert-0',
          'hover:opacity-5',
          'hover:saturate-150',
          'hover:text-base',
          'md:font-semibold',
        ],
      },
    ]);
  });

  it('should consider `prefix`, `separator` and `corePlugins` configurations', async () => {
    const converter = createTailwindConverter({
      tailwindConfig: {
        content: [],
        prefix: 'tw-',
        separator: '_',
        corePlugins: {
          fontWeight: false,
          borderColor: false,
        },
      },
    });
    const converted = await converter.convertCSS(simpleCSS);

    expect(converted.convertedRoot.toString()).toMatchSnapshot();
    expect(converted.nodes).toEqual([
      {
        rule: expect.objectContaining({ selector: '.foo' }),
        tailwindClasses: [
          'tw-text-xs',
          'tw-py-3',
          'tw-border-r-2',
          'tw-border-dashed',
          'hover_tw-blur-sm',
          'hover_tw-brightness-50',
          'hover_tw-sepia',
          'hover_tw-contrast-100',
          'hover_tw-hue-rotate-30',
          'hover_tw-invert-0',
          'hover_tw-opacity-5',
          'hover_tw-saturate-150',
          'hover_tw-text-base',
        ],
      },
    ]);
  });

  it('should convert unconvertible declarations if `arbitraryPropertiesIsEnabled` config is enabled', async () => {
    const converter = createTailwindConverter({
      arbitraryPropertiesIsEnabled: true,
      tailwindConfig: {
        content: [],
        corePlugins: {
          borderColor: false,
        },
      },
    });
    const converted = await converter.convertCSS(simpleCSS);

    expect(converted.convertedRoot.toString()).toMatchSnapshot();
    expect(converted.nodes).toEqual([
      {
        rule: expect.objectContaining({ selector: '.foo' }),
        tailwindClasses: [
          'text-xs',
          '[animation-delay:200ms]',
          '[border:4px_solid_transparent]',
          'py-3',
          'border-r-2',
          'border-dashed',
          'hover:blur-sm',
          'hover:brightness-50',
          'hover:sepia',
          'hover:contrast-100',
          'hover:hue-rotate-30',
          'hover:invert-0',
          'hover:opacity-5',
          'hover:saturate-150',
          'hover:[transform:translateX(12px)_translateY(0.5em)_translateZ(0.5rem)_scaleY(0.725)_rotate(124deg)]',
          'hover:text-base',
          'md:font-semibold',
        ],
      },
    ]);
  });

  it('should not prefix arbitrary properties', async () => {
    const converter = createTailwindConverter({
      arbitraryPropertiesIsEnabled: true,
      tailwindConfig: {
        content: [],
        prefix: 'tw-',
        separator: '_',
        corePlugins: {
          borderColor: false,
        },
      },
    });
    const converted = await converter.convertCSS(simpleCSS);

    expect(converted.convertedRoot.toString()).toMatchSnapshot();
    expect(converted.nodes).toEqual([
      {
        rule: expect.objectContaining({ selector: '.foo' }),
        tailwindClasses: [
          'tw-text-xs',
          '[animation-delay:200ms]',
          '[border:4px_solid_transparent]',
          'tw-py-3',
          'tw-border-r-2',
          'tw-border-dashed',
          'hover_tw-blur-sm',
          'hover_tw-brightness-50',
          'hover_tw-sepia',
          'hover_tw-contrast-100',
          'hover_tw-hue-rotate-30',
          'hover_tw-invert-0',
          'hover_tw-opacity-5',
          'hover_tw-saturate-150',
          'hover_[transform:translateX(12px)_translateY(0.5em)_translateZ(0.5rem)_scaleY(0.725)_rotate(124deg)]',
          'hover_tw-text-base',
          'md_tw-font-semibold',
        ],
      },
    ]);
  });

  it('should return an empty result when converting an empty string', async () => {
    const converter = createTailwindConverter();
    const converted = await converter.convertCSS('');

    expect(converted.convertedRoot.toString()).toEqual('');
    expect(converted.nodes).toEqual([]);
  });

  it('should convert the css part string', async () => {
    const converter = createTailwindConverter();
    const converted = await converter.convertCSS(
      '{ text-align: center; font-size: 12px; &:hover { font-size: 16px; } @media screen and (min-width: 768px) { font-weight: 600; } }',
    );
    expect(converted.convertedRoot.toString()).toMatchSnapshot();
    expect(converted.nodes).toEqual([
      expect.objectContaining({
        rule: expect.objectContaining({ selector: '' }),
        tailwindClasses: ['text-center', 'text-xs', 'hover:text-base', 'md:font-semibold'],
      }),
    ]);
  });

  it('should throw an error when converting invalid css string', async () => {
    const converter = createTailwindConverter();
    await expect(
      converter.convertCSS('some invalid css string... .some-class { display: block; } ...'),
    ).rejects.toThrow(Error);
  });

  it('should convert the complex CSS', async () => {
    const converter = createTailwindConverter();
    const converted = await converter.convertCSS(complexCSS);

    expect(converted.convertedRoot.toString()).toMatchSnapshot();
    expect(converted.nodes).toEqual([
      {
        rule: expect.objectContaining({ selector: '.foo' }),
        tailwindClasses: [
          'md:accent-custom-color-gold',
          'animate-spin',
          'aspect-video',
          'content-center',
          'content-end',
          'portrait:text-[black]',
          "after:content-['*']",
          'after:align-text-top',
          'after:origin-top',
          'hidden:delay-150',
          'hidden:duration-200',
          'hidden:transition',
          'hidden:ease-in',
        ],
      },
      {
        rule: expect.objectContaining({ selector: '.foo .baz' }),
        tailwindClasses: [
          'md:text-center',
          'place-content-around',
          'place-items-center',
          'place-self-stretch',
          'pointer-events-auto',
          'relative',
          'resize-x',
          '-right-32',
          'lg:backdrop-brightness-90',
          'lg:backdrop-sepia-[25%]',
          'lg:backdrop-blur-none',
          'motion-safe:custom-screen:supports-flex:order-[-123]',
          'motion-safe:custom-screen:supports-flex:tracking-[0.25rem]',
          'motion-safe:custom-screen:supports-flex:leading-snug',
          'motion-safe:custom-screen:supports-flex:list-inside',
          'motion-safe:custom-screen:supports-flex:list-decimal',
          'motion-safe:custom-screen:supports-flex:mb-[-0.875rem]',
          'motion-safe:custom-screen:supports-flex:max-h-full',
          'motion-safe:custom-screen:supports-flex:max-w-screen-2xl',
          'motion-safe:custom-screen:supports-flex:min-h-fit',
          'motion-safe:custom-screen:supports-flex:min-w-min',
          'motion-safe:custom-screen:supports-flex:mix-blend-color-dodge',
          'motion-safe:custom-screen:supports-flex:object-fill',
          'motion-safe:custom-screen:supports-flex:object-right-top',
          'motion-safe:custom-screen:supports-flex:ml-[2em]',
          'motion-safe:custom-screen:supports-flex:mr-[1vh]',
          'motion-safe:custom-screen:supports-flex:mt-3',
          'motion-safe:custom-screen:supports-flex:mt-[3vw]',
          'motion-safe:custom-screen:supports-flex:-mb-2.5',
          'motion-safe:custom-screen:supports-flex:mx-6',
          'motion-safe:custom-screen:supports-flex:left-2',
          'supports-[scroll-snap-align:end]:snap-end',
          'supports-[scroll-snap-align:end]:snap-always',
          'supports-[scroll-snap-align:end]:line-through',
          'supports-[scroll-snap-align:end]:scroll-mt-[12%]',
          'supports-[scroll-snap-align:end]:scroll-pl-3.5',
          'supports-[scroll-snap-align:end]:scroll-pr-[10vw]',
          'supports-[scroll-snap-align:end]:scroll-pt-[10em]',
          'supports-[scroll-snap-align:end]:scroll-pb-5',
          'supports-[scroll-snap-align:end]:scroll-p-[100px]',
        ],
      },
      {
        rule: expect.objectContaining({ selector: '.bar' }),
        tailwindClasses: [
          'md:content-center',
          'md:items-start',
          'lg:backdrop-brightness-75',
          'lg:backdrop-sepia',
          'lg:bg-local',
          'lg:content-end',
          'lg:items-center',
          'animate-[some-animation_2s_linear_infinite]',
          'origin-[12%_25.5%]',
          'ease-[cubic-bezier(0.23,0,0.25,1)]',
          'lg:bg-blend-difference',
          'lg:bg-clip-padding',
          'lg:bg-[hsl(30,51%,22%)]',
          'lg:disabled:bg-gradient-to-tr',
          'lg:disabled:bg-origin-padding',
          'lg:disabled:bg-left-bottom',
          'lg:disabled:bg-no-repeat',
          'lg:disabled:bg-contain',
          'lg:disabled:border-b-2',
          'after:border-spacing-[5%]',
          'after:rounded-full',
          'after:border-l-[3px]',
          'after:border-l-transparent',
          'after:border-l-[1rem]',
          'after:border-r-2',
          'after:border-r-[aqua]',
          'after:border-r-8',
          'after:border-dashed',
          'after:border-t-[some-invalid-value]',
          'after:flex-1',
          'after:rounded-tl-none',
          'after:rounded-tr-[0.25%]',
          'after:border-t-[current]',
          'after:border-t-[100vh]',
          'after:border-dotted',
          'after:border-0',
          'after:bottom-[100vw]',
          'box-decoration-slice',
          'shadow',
          'box-border',
          'break-after-all',
          'break-before-page',
          'break-inside-avoid-column',
          'caret-[color:var(--cyan)]',
          'h-9',
          'flex-1',
          'xl:flex-[1_0]',
          'xl:clear-both',
          'xl:text-lime-200',
          'xl:gap-x-48',
          'xl:columns-3',
          'xl:content-none',
          'xl:cursor-pointer',
          'xl:hidden',
          'xl:fill-sky-800',
          'xl:blur-sm',
          'xl:brightness-50',
          'xl:sepia',
          'xl:contrast-100',
          'xl:hue-rotate-30',
          'xl:invert-0',
          'xl:opacity-5',
          'xl:saturate-150',
          'xl:flex-auto',
          'xl:basis-3',
          'xl:flex-col-reverse',
          'xl:grow',
          'xl:shrink-0',
          'xl:flex-wrap-reverse',
          'xl:float-right',
          'xl:text-2xl',
          'xl:antialiased',
          'xl:italic',
          'xl:ordinal',
          'xl:font-semibold',
        ],
      },
      {
        rule: expect.objectContaining({ selector: '.foo' }),
        tailwindClasses: ['appearance-none', 'disabled:opacity-0', 'disabled:invisible', 'disabled:select-none'],
      },
      {
        rule: expect.objectContaining({ selector: '.foo[some-attribute]' }),
        tailwindClasses: ['select-text'],
      },
      {
        rule: expect.objectContaining({
          selector: ".foo [aria-role='button']",
        }),
        tailwindClasses: [
          'uppercase',
          'underline-offset-[1rem]',
          'touch-pan-left',
          'origin-bottom-right',
          'transition-none',
        ],
      },
      {
        rule: expect.objectContaining({
          selector: ".foo[aria-hidden='false']",
        }),
        tailwindClasses: [
          'collapse',
          'whitespace-pre-line',
          'w-6/12',
          'will-change-transform',
          'break-all',
          'z-40',
          'translate-x-3',
          'translate-y-[-0.5em]',
          'skew-x-1',
          'skew-y-3',
          'rotate-[-0.25turn]',
          'transition-colors',
          'duration-200',
          'ease-out',
          '-scale-x-75',
          'scale-y-105',
        ],
      },
      {
        rule: expect.objectContaining({ selector: '.foo .bar' }),
        tailwindClasses: [
          'translate-x-[10px_0.625rem]',
          'skew-x-2',
          '-rotate-45',
          'pl-[12%]',
          'pr-[100vw]',
          'pt-64',
          'pb-1',
          'px-6',
          'py-8',
          '-scale-75',
        ],
      },
      {
        rule: expect.objectContaining({ selector: '.foo .baz > .foo-bar' }),
        tailwindClasses: [
          'text-left',
          'text-[length:var(--some-size)]',
          'text-[color:var(--some-color)]',
          'active:focus:break-after-avoid',
          'active:focus:break-before-left',
          'active:focus:break-inside-auto',
          'xl:isolate',
          'xl:justify-center',
          'xl:active:text-sky-800',
          'xl:active:focus:justify-items-start',
          'xl:active:focus:justify-self-end',
          'motion-safe:custom-screen:supports-flex:opacity-20',
          'motion-safe:custom-screen:supports-flex:-order-last',
          'motion-safe:custom-screen:supports-flex:outline-lime-600',
          'motion-safe:custom-screen:supports-flex:outline-offset-2',
          'motion-safe:custom-screen:supports-flex:outline-dotted',
          'motion-safe:custom-screen:supports-flex:outline-2',
          'motion-safe:custom-screen:supports-flex:overflow-hidden',
          'motion-safe:custom-screen:supports-flex:break-words',
          'motion-safe:custom-screen:supports-flex:overflow-x-scroll',
          'motion-safe:custom-screen:supports-flex:overflow-y-visible',
          'motion-safe:custom-screen:supports-flex:overscroll-contain',
          'motion-safe:custom-screen:supports-flex:overscroll-x-auto',
          'motion-safe:custom-screen:supports-flex:overscroll-y-none',
        ],
      },
      {
        rule: expect.objectContaining({ selector: '.foo div > [data-zoo]' }),
        tailwindClasses: ['border', 'pl-[25%]', 'pr-[1.5em]', 'pt-0', 'pb-px', 'bottom-full'],
      },
      {
        rule: expect.objectContaining({ selector: '.loving .bar > .testing' }),
        tailwindClasses: [
          "lg:bg-[url('/some-path/to/large\\_image.jpg')]",
          'lg:border-custom-color-gold',
          'lg:border-b-custom-color-200',
          'lg:border-b-[length:var(--some-size)]',
          'lg:border-neutral-600',
          'lg:border-t-custom-color-400',
          'lg:rounded-br-sm',
          'lg:rounded-bl',
          'lg:border-b-[2em]',
          'lg:border-b-[#ff0000]',
          'lg:border-4',
          'lg:border-solid',
          'lg:border-dashed',
          'lg:border-separate',
        ],
      },
      {
        rule: expect.objectContaining({ selector: '.foo .baz' }),
        tailwindClasses: [
          'gap-[19px]',
          'auto-cols-min',
          'grid-flow-row',
          'auto-rows-max',
          'col-span-3',
          'col-end-4',
          'gap-x-12',
          'col-start-3',
        ],
      },
      {
        rule: expect.objectContaining({ selector: '.foo .baz > .foo-bar' }),
        tailwindClasses: [
          'gap-8',
          'row-span-full',
          'row-end-2',
          'gap-y-6',
          'row-start-auto',
          'grid-cols-2',
          'grid-rows-5',
          'h-[$some-invalid]',
          'active:text-[red]',
        ],
      },
      {
        rule: expect.objectContaining({ selector: '#some-id' }),
        tailwindClasses: [
          'opacity-40',
          'order-last',
          'outline-teal-900',
          'outline-offset-2',
          'outline-8',
          'supports-[display:block]:gap-y-80',
          'supports-[display:block]:scroll-smooth',
          'supports-[display:block]:scroll-ml-2',
          'supports-[display:block]:scroll-mr-[1.5em]',
          'supports-[display:block]:scroll-mb-8',
          'supports-[display:block]:scroll-m-40',
        ],
      },
      {
        rule: expect.objectContaining({ selector: 'div > [data-zoo]' }),
        tailwindClasses: [
          'stroke-[black]',
          'stroke-2',
          'table-fixed',
          'text-justify',
          'decoration-custom-color-gold',
          'line-through',
          'decoration-dotted',
          'decoration-8',
          'indent-[0.125rem]',
          'text-ellipsis',
        ],
      },
    ]);
  });
});
