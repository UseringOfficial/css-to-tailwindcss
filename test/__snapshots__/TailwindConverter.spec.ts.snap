// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TailwindConverter > should consider \`prefix\`, \`separator\` and \`corePlugins\` configurations 1`] = `
".foo {
  @apply tw-pt-3 tw-pb-3 tw-text-xs tw-border-r-2 tw-border-dashed hover_tw-blur-sm hover_tw-brightness-50 hover_tw-sepia hover_tw-contrast-100 hover_tw-hue-rotate-30 hover_tw-invert-0 hover_tw-opacity-5 hover_tw-saturate-150 hover_tw-text-base;
  animation-delay: 200ms;
  border: 4px solid transparent;
}
.foo:hover {
  transform: translateX(12px) translateY(0.5em) translateZ(0.5rem)
      scaleY(0.725) rotate(124deg);
}
@media screen and (min-width: 768px) {
  .foo {
    font-weight: 600
  }
}
@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
"
`;

exports[`TailwindConverter > should convert the complex CSS 1`] = `
"@charset "UTF-8";
.foo{
  @apply md:accent-custom-color-gold animate-spin aspect-video content-end portrait:text-[black] after:content-['*'] after:align-text-top after:origin-top hidden:delay-150 hidden:duration-200 hidden:transition hidden:ease-in;
}
.foo .baz{
  @apply md:text-center place-content-around place-items-center place-self-stretch pointer-events-auto relative resize-x -right-32 lg:backdrop-brightness-90 lg:backdrop-sepia-[25%] lg:backdrop-blur-0 motion-safe:custom-screen:supports-flex:order-[-123] motion-safe:custom-screen:supports-flex:left-2 motion-safe:custom-screen:supports-flex:tracking-[0.25rem] motion-safe:custom-screen:supports-flex:leading-snug motion-safe:custom-screen:supports-flex:list-inside motion-safe:custom-screen:supports-flex:list-decimal motion-safe:custom-screen:supports-flex:-mb-2.5 motion-safe:custom-screen:supports-flex:ml-[2em] motion-safe:custom-screen:supports-flex:mr-[1vh] motion-safe:custom-screen:supports-flex:mt-[3vw] motion-safe:custom-screen:supports-flex:max-h-full motion-safe:custom-screen:supports-flex:max-w-screen-2xl motion-safe:custom-screen:supports-flex:min-h-fit motion-safe:custom-screen:supports-flex:min-w-min motion-safe:custom-screen:supports-flex:mix-blend-color-dodge motion-safe:custom-screen:supports-flex:object-fill motion-safe:custom-screen:supports-flex:object-right-top supports-[scroll-snap-align:end]:scroll-mt-[12%] supports-[scroll-snap-align:end]:scroll-pb-5 supports-[scroll-snap-align:end]:scroll-pl-3.5 supports-[scroll-snap-align:end]:scroll-pr-[10vw] supports-[scroll-snap-align:end]:scroll-pt-[10em] supports-[scroll-snap-align:end]:snap-end supports-[scroll-snap-align:end]:snap-always supports-[scroll-snap-align:end]:line-through;
}
.bar{
  @apply md:content-center md:items-start lg:backdrop-brightness-75 lg:backdrop-sepia lg:bg-local lg:content-end lg:items-center animate-[some-animation_2s_linear_infinite] origin-[12%_25.5%] ease-[cubic-bezier(0.23,0,0.25,1)] lg:bg-blend-difference lg:bg-clip-padding lg:bg-[hsl(30,51%,22%)] lg:disabled:bg-gradient-to-tr lg:disabled:bg-origin-padding lg:disabled:bg-left-bottom lg:disabled:bg-no-repeat lg:disabled:bg-contain lg:disabled:border-b-2 after:border-dashed after:border-l-transparent after:border-l-[1rem] after:rounded-full after:border-r-[aqua] after:border-r-8 after:border-spacing-[5%] after:border-dotted after:border-t-[current] after:rounded-tl-none after:rounded-tr-[0.25%] after:border-0 after:bottom-[100vw] after:flex-1 box-decoration-slice shadow box-border break-after-all break-before-page break-inside-avoid-column caret-[color:var(--cyan)] h-9 flex-1 xl:flex-[1_0] xl:clear-both xl:text-lime-200 xl:gap-x-48 xl:columns-3 xl:content-none xl:cursor-pointer xl:hidden xl:fill-sky-800 xl:blur-sm xl:brightness-50 xl:sepia xl:contrast-100 xl:hue-rotate-30 xl:invert-0 xl:opacity-5 xl:saturate-150 xl:flex-auto xl:basis-3 xl:flex-col-reverse xl:grow xl:shrink-0 xl:flex-wrap-reverse xl:float-right xl:text-2xl xl:antialiased xl:italic xl:ordinal xl:font-semibold;
}
@media (min-width: 768px) {

  /* custom-color-gold */
}
@media (min-width: 769px) {
  .foo {
    @apply appearance-none disabled:opacity-0 disabled:invisible disabled:select-none;
  }
  .foo[some-attribute] {
    @apply select-text;
  }
}
.foo {
  --some-size: 12px;
  --some-color: #333333;
  align-self: self-end;

  /* unconvertable value */

  /* duplicate */
}

/* duplicate with another value */
.foo [aria-role='button'] {
  @apply uppercase underline-offset-[1rem] touch-pan-left origin-bottom-right transition-none;

  /* unconvertable aria selector */
  transform: translateX(12px) translateY(0.5em) translateZ(0.5rem)
      scaleY(0.725) rotate(124deg);
}
.foo[aria-hidden='true'] {
  transform: some invalid-transform;
}
.foo[aria-hidden='false'] {
  @apply collapse whitespace-pre-line w-6/12 will-change-transform break-all z-40 translate-x-3 translate-y-[-0.5em] skew-x-1 skew-y-3 -scale-x-75 scale-y-105 rotate-[-0.25turn] transition-colors duration-200 ease-out;

  /* unconvertable aria selector */
}
.foo .bar {
  @apply pb-1 pl-[12%] pr-[100vw] pt-64 translate-x-[10px_0.625rem] skew-x-2 -scale-x-75 -scale-y-75 -rotate-45;
  transition: transform 200ms cubic-bezier(0, 0, 0.2, 1), color,
      background-color, border-color, text-decoration-color, fill,
      stroke 200ms cubic-bezier(0.4, 0, 0.2, 1);
}
.foo .baz > .foo-bar {
  @apply text-left text-[length:var(--some-size)] text-[color:var(--some-color)] active:focus:break-after-avoid active:focus:break-before-left active:focus:break-inside-auto xl:isolate xl:justify-center xl:active:text-sky-800 xl:active:focus:justify-items-start xl:active:focus:justify-self-end motion-safe:custom-screen:supports-flex:opacity-20 motion-safe:custom-screen:supports-flex:-order-last motion-safe:custom-screen:supports-flex:outline-lime-600 motion-safe:custom-screen:supports-flex:outline-offset-2 motion-safe:custom-screen:supports-flex:outline-dotted motion-safe:custom-screen:supports-flex:outline-2 motion-safe:custom-screen:supports-flex:overflow-hidden motion-safe:custom-screen:supports-flex:break-words motion-safe:custom-screen:supports-flex:overflow-x-scroll motion-safe:custom-screen:supports-flex:overflow-y-visible motion-safe:custom-screen:supports-flex:overscroll-contain motion-safe:custom-screen:supports-flex:overscroll-x-auto motion-safe:custom-screen:supports-flex:overscroll-y-none;
}
.foo div > [data-zoo] {
  @apply pt-0 pr-[1.5em] pb-px pl-[25%] border bottom-full;
}
.bar {

  /* arbitary value */
  animation-delay: 200ms;

  /* unconvertable value */
}
.loving .bar > .testing{
  @apply lg:bg-[url('/some-path/to/large\\_image.jpg')] lg:border-4 lg:border-dashed lg:rounded-bl lg:rounded-br-sm lg:border-b-[length:var(--some-size)] lg:border-separate lg:border-neutral-600 lg:border-t-custom-color-400;
}
@media screen and (min-width: 1024px) {
  .bar[aria-disabled='true'] {
    background-attachment: initial;
  }
  .loving .bar > .testing {

    /* custom-color-gold */

    /* custom-color-200 */

    /* neutral-600 */
  }
}
.bar {
  --cyan: #06b6d4;
}
@media (min-width: 1280px) {
  .bar {

    /* sky-800 */
  }
}
@media screen and (min-width: 1280px) {
  .bar {
    flex-flow: column;
  }
}
@media screen and (min-width: 1280px) and (max-width: 1440px) {
  .foo .baz {
    @apply gap-[19px] auto-cols-min grid-flow-row auto-rows-max col-span-3 col-end-4 gap-x-12 col-start-3;
    grid: 'a' 200px 'b' max-content;
  }
  .foo .baz > .foo-bar {
    @apply gap-8 row-span-full row-end-2 gap-y-6 row-start-auto grid-cols-2 grid-rows-5 h-[$some-invalid] active:text-[red];
  }
}
@supports (display: flex) {
  @media (min-width: 768px) {
    @media (prefers-reduced-motion: no-preference) {
      @media (max-width: 1024px) {
        .foo .baz > .foo-bar {
          outline: 4px dashed rgb(254, 240, 138);

          /* yellow-200 */

          /* lime-600 */
        }
      }
    }
  }
}
#some-id {
  @apply opacity-40 order-last outline-teal-900 outline-offset-2 outline-8 supports-[display:block]:gap-y-80 supports-[display:block]:scroll-smooth supports-[display:block]:scroll-mt-40 supports-[display:block]:scroll-mb-8 supports-[display:block]:scroll-ml-2 supports-[display:block]:scroll-mr-[1.5em];
  outline: none;

  /* teal-900 */
  outline-style: groove;
}
@supports (scroll-snap-align: end) {
  .foo .baz {
    scroll-snap-type: x mandatory;
  }
}
div > [data-zoo] {
  @apply stroke-[black] stroke-2 table-fixed text-justify decoration-custom-color-gold line-through decoration-dotted decoration-8 indent-[0.125rem] text-ellipsis;
  text-decoration: underline #123456 dashed 4px;

  /* custom-color-100 */

  /* custom-color-gold */
}
"
`;

exports[`TailwindConverter > should convert the css part string 1`] = `
"{
    @apply text-center text-xs hover:text-base md:font-semibold;
}"
`;

exports[`TailwindConverter > should convert the simple CSS 1`] = `
".foo {
  @apply pt-3 pb-3 text-xs border-4 border-solid border-transparent hover:blur-sm hover:brightness-50 hover:sepia hover:contrast-100 hover:hue-rotate-30 hover:invert-0 hover:opacity-5 hover:saturate-150 hover:text-base md:font-semibold;
  animation-delay: 200ms;
}
.foo:hover {
  transform: translateX(12px) translateY(0.5em) translateZ(0.5rem)
      scaleY(0.725) rotate(124deg);
}
@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
"
`;

exports[`TailwindConverter > should convert unconvertible declarations if \`arbitraryPropertiesIsEnabled\` config is enabled 1`] = `
".foo {
    @apply pt-3 pb-3 text-xs [animation-delay:200ms] border-r-2 border-dashed [border:4px_solid_transparent] hover:blur-sm hover:brightness-50 hover:sepia hover:contrast-100 hover:hue-rotate-30 hover:invert-0 hover:opacity-5 hover:saturate-150 hover:[transform:translateX(12px)_translateY(0.5em)_translateZ(0.5rem)_scaleY(0.725)_rotate(124deg)] hover:text-base md:font-semibold;
}
@keyframes spin {
    0% {
        transform: rotate(0deg);
    }
    100% {
        transform: rotate(360deg);
    }
}
"
`;

exports[`TailwindConverter > should not prefix arbitrary properties 1`] = `
".foo {
    @apply tw-pt-3 tw-pb-3 tw-text-xs [animation-delay:200ms] tw-border-r-2 tw-border-dashed [border:4px_solid_transparent] hover_tw-blur-sm hover_tw-brightness-50 hover_tw-sepia hover_tw-contrast-100 hover_tw-hue-rotate-30 hover_tw-invert-0 hover_tw-opacity-5 hover_tw-saturate-150 hover_[transform:translateX(12px)_translateY(0.5em)_translateZ(0.5rem)_scaleY(0.725)_rotate(124deg)] hover_tw-text-base md_tw-font-semibold;
}
@keyframes spin {
    0% {
        transform: rotate(0deg);
    }
    100% {
        transform: rotate(360deg);
    }
}
"
`;
